Paco/src/com/pacoapp/paco/model/ExperimentProviderUtil.java

```diff
@@ -814,35 +811,54 @@ public void insertEvent(EventInterface eventI) {
       String colNameConcat = (groupBy != null) ? criteriaColumns.concat(BLANK).concat(groupBy) : criteriaColumns;
       colNameConcat = (groupBy != null && having != null) ? colNameConcat.concat(BLANK).concat(having) : colNameConcat;
+      colNameConcat = (sortOrder != null)?colNameConcat.concat(BLANK).concat(sortOrder) : colNameConcat;
       allColumns.addAll(ExperimentUtil.aggregateExtractedColNames(colNameConcat));
-      String tableIndicator = ExperimentUtil.identifyTablesInvolved(eventsOutputColumns, allColumns);
-      if (tableIndicator.equals(ExperimentProvider.EVENTS_OUTPUTS_TABLE_NAME)) {
-        cursor = dbHelper.query(ExperimentProvider.EVENTS_OUTPUTS_DATATYPE, projection, criteriaColumns, criteriaValues,
+
+      if (sortOrder != null && limit != null) {
+        sortOrder = sortOrder.concat(LIMIT).concat(limit);
+      } 
+      
+      //identify all tables involved for the complete column names
+      String tableIndicator = ExperimentUtil.identifyTablesInvolved(eventsColumns, allColumns);
+      
+      //adding a default projection of event table primary key column
+      int crtLength = projection.length ;
+      modifiedProjection = new String[crtLength+1];
+      System.arraycopy(projection, 0, modifiedProjection, 0, crtLength);
+      //adding the following columns in the projection list to help in coalescing
+      modifiedProjection[crtLength]="EVENTS._ID";
+      
+      if (tableIndicator.equals(ExperimentProvider.EVENTS_OUTPUTS_TABLE_NAME)) { 
+        cursor = dbHelper.query(ExperimentProvider.EVENTS_OUTPUTS_DATATYPE, modifiedProjection, criteriaColumns, criteriaValues,
                                 sortOrder, groupBy, having);
       } else if (tableIndicator.equals(ExperimentProvider.EVENTS_TABLE_NAME)) {
-        cursor = dbHelper.query(ExperimentProvider.EVENTS_DATATYPE, projection, criteriaColumns, criteriaValues,
-                                sortOrder, groupBy, having);
-      } else if (tableIndicator.equals(ExperimentProvider.OUTPUTS_TABLE_NAME)) {
-        cursor = dbHelper.query(ExperimentProvider.OUTPUTS_DATATYPE, projection, criteriaColumns, criteriaValues,
+        cursor = dbHelper.query(ExperimentProvider.EVENTS_DATATYPE, modifiedProjection, criteriaColumns, criteriaValues,
                                 sortOrder, groupBy, having);
-      }
-
+      } 
+      //to maintain the insertion order
+      eventMap = Maps.newLinkedHashMap();
+    
       if (cursor != null) {
         events = Lists.newArrayList();
         while (cursor.moveToNext()) {
-          Event event = ExperimentUtil.createEventWithPartialResponses(cursor);
-          events.add(event);
+          //if no need to coalesce, we just add it to the list and send the collection to the client.
+          event = createEvent(cursor, false);
+          Event oldEvent = eventMap.get(event.getId()); 
+          if(oldEvent == null){
+            event.setResponses(findResponsesFor(event));
```

Comment:
The first time we come across an event, we find all responses for that event, and add it to the event. This complete event is what we store in map. So, in the subsequent times, when we encounter the same event, we do not need to do anything.